File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.py
import sys

sys.setrecursionlimit(10**9)
input_list = []

while True:
    try:
        input_list.append(int(sys.stdin.readline()))
    except:
        break

def pre_to_post(preordered):
    if not preordered:
        return []
    else:
        root = preordered[0]
        lt_subtree = []
        rt_subtree = []
        for val in preordered:
            if val == root:
                pass
            elif val < root:
                lt_subtree.append(val)
            else:
                rt_subtree.append(val)
            
        return pre_to_post(lt_subtree) + pre_to_post(rt_subtree) + [root]

for node in pre_to_post(input_list):
    print(node)

File: 백준/Gold/5639. 이진 검색 트리/이진 검색 트리.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String s = "";
        ArrayList<Integer> inputList = new ArrayList<>();
        while ((s = br.readLine()) != null) {
            inputList.add(Integer.parseInt(s));
        }

        ArrayList<Integer> postOrdered = preToPost(inputList);
        for (int val : postOrdered) {
            bw.write(val + "\n");
        }

        bw.flush();

        bw.close();
        br.close();
    }

    static ArrayList<Integer> preToPost(ArrayList<Integer> preordered) {
        if (preordered.isEmpty()) {
            return new ArrayList<>();
        }
        int root = preordered.get(0);

        ArrayList<Integer> ltSubtree = new ArrayList<>();
        ArrayList<Integer> rtSubtree = new ArrayList<>();

        for (int val : preordered) {
            if (val < root) {
                ltSubtree.add(val);
            } else if (val > root) {
                rtSubtree.add(val);
            }
        }
        ArrayList<Integer> result = new ArrayList<>();
        result.addAll(preToPost(ltSubtree));
        result.addAll(preToPost(rtSubtree));
        result.add(root);

        return result;
    }
}